src/ucp/proto/proto_am.inl

```diff
@@ -24,6 +24,48 @@
 typedef void (*ucp_req_complete_func_t)(ucp_request_t *req, ucs_status_t status);
 
 
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_copy_header_to_internal_buf(ucp_request_t *req)
+{
+    void *user_header;
+    unsigned flags_mask = (UCP_AM_SEND_FLAG_COPY_HEADER | 
+                           UCP_AM_SEND_PRIV_FLAG_HEADER_COPIED);
+    unsigned flags = req->send.msg_proto.am.flags & flags_mask;
+    
+    if ((req->send.msg_proto.am.header_length == 0) ||
+        (flags != UCP_AM_SEND_FLAG_COPY_HEADER)) {
```

Comment:
Checking "&" is maybe not enough. We need to check also that we didn't already copy the header. Not sure if this scenario is possible (Maybe pushing pending request back to pending queue?), But still I think it's better to go safe.

Regarding this if statement, I thought to replace the order, first checking the flags and then checking the length. It's can be better because if the App is not using the flag it's probably do it for more than one message...

Reply:
Pls do things explicitly, it improves code readability. If you want to highlight "impossible" path/state, add an assertion.
IMO calculations at 31-33 lines are redundant but if delete it, adding new flag will break this logic.. Try to make different pieces of code less dependent on each other.

Reply:
It's a possible path from what I see...